core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,303 @@
+          buf = ByteBuffer.allocate(nextMessageSize);
+          while (buf.remaining() > 0) {
+            channel.read(buf);
+          }
+          queues.get(fromPartyId).add(buf.array());
+        } catch (IOException e) {
+          running = false;
+        }
+      }
+    }
+
+  }
+
+  private class Handshaker implements Runnable {
+    private ServerSocketChannel server;
+    private Map<Integer, SocketChannel> partyIdToChannel;
+    private Semaphore blocker;
+    private ConcurrentLinkedDeque<Boolean> successes;
+
+    public Handshaker(ServerSocketChannel server, Map<Integer, SocketChannel> partyIdToChannel,
+        Semaphore blocker, ConcurrentLinkedDeque<Boolean> successes) {
+      this.server = server;
+      this.partyIdToChannel = partyIdToChannel;
+      this.blocker = blocker;
+      this.successes = successes;
+    }
+
+    @Override
+    public void run() {
+      int count = 0;
+      while (count < getNoOfParties() - 1) {
+        try {
+          SocketChannel channel = server.accept();
+          channel.configureBlocking(true);
+          ByteBuffer buf = ByteBuffer.allocate(1);
+          channel.read(buf);
+          buf.position(0);
+          int id = buf.get();
+          partyIdToChannel.put(id, channel);
+          successes.add(true);
+          blocker.release();
+          count++;
+        } catch (IOException e) {
+          logger.error("IOException occured during handshake", e);
+          count++;
+          successes.add(false);
+          blocker.release();
+        }
+      }
+    }
+  }
+
+  /**
+   * Communicate with all parties and obtain their ID. This is required since we use a dual
+   * connection strategy which means all parties serve as both client and server. Server just
+   * accepts incoming connections, and we need to map the connection to a party. Malicious parties
+   * could send an incorrect ID, which would overwrite an honest party. This, however, leads to a
+   * null pointer exception as soon as we try to communicate with a missing partyID. Can still lead
+   * to leaking information. Cannot be fixed without authenticated channels.
+   */
+  void handshake() {
+    ConcurrentHashMap<Integer, SocketChannel> partyIdToChannel = new ConcurrentHashMap<>();
+    Semaphore blocker = new Semaphore(-getNoOfParties() + 2);
+    ConcurrentLinkedDeque<Boolean> successes = new ConcurrentLinkedDeque<>();
+    Thread t = new Thread(new Handshaker(this.server, partyIdToChannel, blocker, successes));
+    t.start();
+    for (SocketChannel channel : this.clients.values()) {
+      ByteBuffer id = ByteBuffer.allocate(1);
+      id.put((byte) this.conf.getMyId());
+      id.position(0);
+      ExceptionConverter.safe(() -> channel.write(id), "Not able to send my id");
+    }
+    ExceptionConverter.safe(() -> {
+      blocker.acquire();
+      return null;
+    } , "Blocker got interrupted");
+    boolean connected = true;
+    for (Boolean succ : successes) {
```

Comment:
Couldn't success be signaled from the handshake as a single `boolean`? I do not see the need for this list.